### ISSUE TITLE
update standard library to a better "interface" pattern so that we have a null hypothesis against the interface/oop proposal

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#1829]
Related:
 * #1268 
 * #1214 
 * #591 
 * #1669 

Probably lots more. But the big parent issue is #130. It's one of the biggest topics in the community and we really don't know the direction Zig will go (it's a difficult design decision).

However, one thing we can do to get closer to knowing the answer, is to push the null hypothesis to the best thing it can be. That is, using status quo Zig and accepted proposals, how would we solve the problems that #130 would solve, without any other modifications to the language?

As @Hejsil points out on that issue, the status quo solution to "interfaces" is hostile to optimization. He created a demo of a different pattern here: https://github.com/Hejsil/zig-interface

@tgschultz has done similar prototypes (see for example https://github.com/ziglang/zig/issues/130#issuecomment-447032761) I've also seen him post a bunch of related gists on IRC but I don't have any links.

So it looks like we have a clear way to update the standard library to use this pattern which is at least better on the optimizer. It also fixes a footgun and seems at least as readable, if not more, than status quo. So let's do it in the standard library. Then we can close this issue, and go back to all those other issues, armed with an improved null hypothesis.

**Comment by dbandstra** [comment]
Nice:
- no need for fieldParentPtr (well, traded this for some *c_void evilry)
- create trait binding from concrete implementation, instead of vice versa. The implementation can actually be free of such glue
- don't need to use a pointer to pass the trait objects around (since they are just a box of two pointers). I dunno, it just looks nice

**Comment by Hejsil** [comment]
@dbandstra `*c_void` pointers are harmless as long as you don't `@ptrCast` them (which is just as unsafe as on any other pointer, and this unsafeness is documented). The idea with my `zig-interface` is that we have a function called `vtable.populate` which will populate a `VTable` (any struct with only fn ptr fields) with its implementation. This function will always be safe, as it uses `@typeInfo` to validate that the `VTable` is populated with functions of the correct signature.

So unless the users explicitly shoots thenself with unsafe features, this pattern is safe, unlike interface fields, where a simple copy will break everything.

**Comment by Rocknest** [comment]
Why `*c_void` and not `*@OpaqueType()`?

**Comment by Hejsil** [comment]
@Rocknest No reason really. It was easier to just use `c_void` because it is there. Using `@OpaqueType()` is probably more proper though :)

**Comment by dbandstra** [comment]
I'm wondering, putting aside everything else for the moment, if the compiler/LLVM is able to optimize (remove duplication) in the case where only error sets are different (andrewrk mentioned this in #764), could it also do the same optimization if only a pointer type is different?

If so, couldn't you construct OutStream using the implementation type, instead of just error type?

```
pub fn OutStream(comptime ImplType: type) type {
  return struct {
    const Error = ImplType.WriteError; // assume this exists
```

Now you could use `*ImplType` instead of `*c_void`.

**Comment by dbandstra** [comment]
I've been playing with this and it feels good. It's like interfaces, but the interface is a constructed object instead being some kind of shape that an object follows.

I was able to take advantage of this in a experimental branch I made in my zigutils repo (note: I've locally reimplemented Allocator to Hejsil's vtable idea): https://github.com/dbandstra/zigutils/blob/new-streams/src/Hunk.zig

This is an object which "conforms" to the same interface twice, in two different ways (see the "high" and "low" methods in Hunk). The Hunk object itself doesn't even need to know about the interface(s) it's conforming to (I baked the "low" and "high" methods in, but I didn't have to - they could even live in another file). And since a Hunk (which is a type of allocator) is not able to free, the Hunk struct itself didn't even need `realloc` or `free` methods. Those are added in the interface/trait/adapter/whatever-you-want-to-call-it (HunkSide). So the Hunk struct itself ended up being really clean. It would be perfectly useful by itself (if you didn't need dynamic dispatch) if it weren't for the fact that all the convenient allocation functions are inside Allocator.

Also, see that Hunk has nested trait/whatever implementations: it implements HunkSide, which implements Allocator. That didn't turn out messy at all, and it feels a lot better than the old fieldParentPtr method. And I don't think there are any doubled up indirections either.

I would wait on refactoring std till we've thought more about the error set templatization problem though.

**Comment by tgschultz** [comment]
I've been playing with this and I think I've come up with a pattern that should accomplish our goals with a minimum of API changes from the user perspective. It's a slight modification on Hejsils example which allows us to keep 'convenience' functions as part of the interface instead of separating them out. I've got an example in this [gist](https://gist.github.com/tgschultz/25c05f250e3df6ad34fdfc06bb938661). When I get to the Stream interface I'll have to see what can be done about the ErrorSet (#764) mess there.

**Comment by Hejsil** [comment]
I personally have come to prefer the free function approach more than member functions. What you end up doing, is creating more helper functions yourself and those cannot be member functions unless you do more wrapping. This `Interface` struct is kinda just a hack around the fact that Zig doesn't have [Unified Call Syntax](https://github.com/ziglang/zig/issues/148).

Let structs implement a minimal composable interface, and have free function that implements helpers around this interface.

**Comment by tgschultz** [comment]
I don't necessarily disagree. However doing it this way incurs a minimal API change from the user perspective, and while big API changes aren't really a big deal pre-1.0.0, I personally think that one should be its own issue and PR since it isn't necessary to accomplish the goals of this one.

Doing it this way also allows for a relatively straight-forward trait implementation re #1669, if we decide to go that route, for example by comparing the parts of the type name or checking for the existence and value of a pub const.

**Comment by Hejsil** [comment]
@tgschultz Alright, reasonable :+1:

**Comment by kristate** [comment]
I am more for @Hejsil's approach at this point. I think it is cleaner and easier to document.

**Comment by tgschultz** [comment]
I'm giving this another go with what I learned from the previous effort. This iteration of the pattern can be seen in [interface.zig](https://github.com/tgschultz/zig/blob/f4099dbac8078ec000cd672273d57e2e27a9541e/std/interface.zig), which contains some helper types and functions. 

Improvements of this over the previous iteration:
- implementation pointers can be null (c_allocator, NullOutStream).
- interfaces are only one struct instead of two.
- no ugly hacks should be necessary to work with current async.

Changes:
- A global vtable is eschewed in favor of storing function pointers directly. This is mostly so that no ugly hacks are required to work with async.

Regressions:
- Using an interface over a comptime-known implementation no longer supports the `self.method()` syntax and instead requires `Interface.method(self)`, but abstracted/runtime interfaces still support `self.method()` syntax.

Notes (both iterations):
- Interfaces with limited error sets (Allocator) will prefer to use the abstracted/runtime version because it can be stored without a generic struct and testing on godbolt indicates it optimizes well anyway.
- Interfaces using highly variable error sets (*Stream) will prefer the comtime-known version since it preserves error sets without the need to pass them around manually.
- Interfaces now contain only pointers and are therefore completely safe to copy.
- Interface function implementations can no longer be of arbitrarily named.

**Comment by tgschultz** [comment]
So this second crack at this isn't going so well. The first issue is that `async<>` has proven more stubborn than anticipated, but a more important problem is that type erasure doesn't seem to work at comptime, the compiler is too aware of the actual types. I probably would have ran into this in the first iteration as well except that I never touched std.rand, which is used as part of the build process.

**Comment by Rocknest** [comment]
I had an idea recently that `@reify` (#383) or something similar could be used to solve this usecase.

```zig
// imaginary code
const Allocator = InterfaceOf(struct {
    alloc: fn (usize) *c_void,
    free: fn (*c_void) void,
});

test "" {
    const impl = SomeAllocator { .a = b };
    const interface = Allocator.init(&iml);
    var ptr = interface.alloc(5);
    interface.free(ptr);
}
```

Function `InterfaceOf` would create a type similar to this one:
```zig
{
  @"$vtable": *struct {...},
  @"$impl": *OpaqueType(),
  pub fn init(impl: var) Allocator {
    // assert impl == Pointer
    return Allocator {
      .@"$vtable" = comptimeGenAllocatorVtable(@typeOf(var).Pointer.Child),
      .@"$impl" = impl,
    };
  }
  pub fn alloc(self: Allocator, arg0: usize) *c_void {
    return self.@"$vtable".alloc(arg0);
  }
  pub fn free(self: Allocator, arg0: *c_void) void {
    self.@"$vtable".free(arg0);
  }
}
```

But im not sure if comptime code generation would be ever allowed in zig.

**Comment by tgschultz** [comment]
My third attempt at this went pretty well. See #2533 for a summary.

**Comment by tgschultz** [comment]
The simpler version I discussed is up now, #2553 . The changes to the pattern over status-quo are that interfaces consist of a pointer to an implementer as well as a pointer to each implementation function. Consequently, instead of the implementer carrying around an interface struct and everything needing a pointer to it, they instead have an `interface()` method (`inStream()`, `outStream()`, `allocator()`, etc) that is used to retrieve the interface struct, which is then free to be copied. A minor concession was made for `async<>` in the form of `Allocator` taking `*const Self` instead of just `Self`.

Rather than use a single `?*@OpaqueType` for all implementation pointers, each interface has its own. Implementers use `Interface.ifaceCast(self)` to assign to the `impl` field of the interface struct, and `interface.implCast(Self)` to retrieve a properly typed pointer to themselves from the interface struct (roughly analogous to `@fieldParentPtr`'s usage in status-quo).

I'm still entirely satisfied with it, as it doesn't really bring much in the way of benefits, but at least it doesn't make things even more complicated than they already are.

It occurred to me during this iteration that we only really care about maintaining ErrorSet information if we intend to switch on an error union response from an interface (or chain of interfaces as in Streams) where the implementer(s) are (all) comptime-known. Consequently, I think I will submit one more variation, which is a modification of #2553 where Streams (the only interface with dynamic ErrorSets currently) are generalized and only return `anyerror`, with some alternative method being employed to construct and cast to the proper ErrorSet in those few situations where it is desired.

@andrewrk: After that, there will be 3 open PRs for consideration. If any of them are considered enough of an improvement over the status-quo to merge, then I'll rebase it after the big posix layer change is merged.

**Comment by tgschultz** [comment]
This variation is identical to the previous one, but I've made the Stream interface ErrorSets non-dynamic. From what I can tell, basically nothing is lost by doing this, some errors just become more standardized versions of themselves: `error.OutOfMemory`,`error.OutOfSpace` => `error.NoSpaceLeft`, etc. Pretty much everything is much cleaner as only like 3 places even took advantage of the dynamic ErrorSets anyway. However I don't think that this is Andrew's vision for how ErrorSets and Streams interact.

From the work on these two I believe I can improve #2533 so it at least works at comptime again and is a bit simpler. However, I will only commit time to that if the consensus is that we want to go in that direction, as it is it should serve well enough to demonstrate the pattern.

Anyways, that it, I'm out. I'm leaving these three PRs (#2533, #2553, #2582) for perusal and consideration. If we decide to move towards one of them let me know and I'll rebase it.

**Comment by fengb** [comment]
I'm a fan of #2553 and #2582 because it's simpler mental model. With the existing `@fieldParentPtr` pattern, we have to ensure the correct pointer is referenced everywhere, and every accidental struct copy can break without compiler help.

With an embedded opaque interface, we can copy the struct around without problems. This also allows us to return interfaces directly instead of needing to embed them into a wrapper struct.

**Comment by andrewrk** [comment]
#130 has been closed. I don't think there is anything to do here.
